# Security patch for CVE-2026-48779
# Memory exhaustion DoS from tiny fragments and data chunks
# Based on upstream commits bd8756a4, 6019b4e7, and 504a6ef9 (ws 5.2.5)
# Adapted for ws 5.2.4

--- a/lib/receiver.js
+++ b/lib/receiver.js
@@ -26,13 +26,25 @@
    * @param {String} binaryType The type for binary data
    * @param {Object} extensions An object containing the negotiated extensions
    * @param {Number} maxPayload The maximum allowed message length
-   */
-  constructor (binaryType, extensions, maxPayload) {
+   * @param {Number} [maxBufferedChunks=0] The maximum number of
+   *     buffered data chunks
+   * @param {Number} [maxFragments=0] The maximum number of message
+   *     fragments
+   */
+  constructor (
+    binaryType,
+    extensions,
+    maxPayload,
+    maxBufferedChunks,
+    maxFragments
+  ) {
     super();
 
     this._binaryType = binaryType || constants.BINARY_TYPES[0];
     this[constants.kWebSocket] = undefined;
     this._extensions = extensions || {};
+    this._maxBufferedChunks = maxBufferedChunks | 0;
+    this._maxFragments = maxFragments | 0;
     this._maxPayload = maxPayload | 0;
 
     this._bufferedBytes = 0;
@@ -48,6 +60,7 @@
 
     this._totalPayloadLength = 0;
     this._messageLength = 0;
+    this._numFragments = 0;
     this._fragments = [];
 
     this._state = GET_INFO;
@@ -64,6 +77,13 @@
   _write (chunk, encoding, cb) {
     if (this._opcode === 0x08) return cb();
 
+    if (
+      this._maxBufferedChunks > 0 &&
+      this._buffers.length >= this._maxBufferedChunks
+    ) {
+      return cb(error(RangeError, 'Too many buffered chunks', false, 1008));
+    }
+
     this._bufferedBytes += chunk.length;
     this._buffers.push(chunk);
     this.startLoop(cb);
@@ -328,6 +348,11 @@
 
     if (this._opcode > 0x07) return this.controlMessage(data);
 
+    if (this._maxFragments > 0 && ++this._numFragments > this._maxFragments) {
+      this._loop = false;
+      return error(RangeError, 'Too many message fragments', false, 1008);
+    }
+
     if (this._compressed) {
       this._state = INFLATING;
       this.decompress(data, cb);
@@ -389,6 +414,7 @@
       this._totalPayloadLength = 0;
       this._messageLength = 0;
       this._fragmented = 0;
+      this._numFragments = 0;
       this._fragments = [];
 
       if (this._opcode === 2) {
--- a/lib/websocket-server.js
+++ b/lib/websocket-server.js
@@ -29,6 +29,10 @@
    * @param {Boolean} options.noServer Enable no server mode
    * @param {Boolean} options.clientTracking Specifies whether or not to track clients
    * @param {(Boolean|Object)} options.perMessageDeflate Enable/disable permessage-deflate
+   * @param {Number} [options.maxBufferedChunks=1048576] The maximum number of
+   *     buffered data chunks
+   * @param {Number} [options.maxFragments=131072] The maximum number of message
+   *     fragments
    * @param {Number} options.maxPayload The maximum allowed message size
    * @param {Function} callback A listener for the `listening` event
    */
@@ -36,6 +40,8 @@
     super();
 
     options = Object.assign({
+      maxBufferedChunks: 1024 * 1024,
+      maxFragments: 128 * 1024,
       maxPayload: 100 * 1024 * 1024,
       perMessageDeflate: false,
       handleProtocols: null,
@@ -287,7 +293,13 @@
     socket.write(headers.concat('\r\n').join('\r\n'));
     socket.removeListener('error', socketOnError);
 
-    ws.setSocket(socket, head, this.options.maxPayload);
+    ws.setSocket(
+      socket,
+      head,
+      this.options.maxPayload,
+      this.options.maxBufferedChunks,
+      this.options.maxFragments
+    );
 
     if (this.clients) {
       this.clients.add(ws);
--- a/lib/websocket.js
+++ b/lib/websocket.js
@@ -114,13 +114,19 @@
    * @param {net.Socket} socket The network socket between the server and client
    * @param {Buffer} head The first packet of the upgraded stream
    * @param {Number} maxPayload The maximum allowed message size
+   * @param {Number} [maxBufferedChunks=0] The maximum number of
+   *     buffered data chunks
+   * @param {Number} [maxFragments=0] The maximum number of message
+   *     fragments
    * @private
    */
-  setSocket (socket, head, maxPayload) {
+  setSocket (socket, head, maxPayload, maxBufferedChunks, maxFragments) {
     const receiver = new Receiver(
       this._binaryType,
       this._extensions,
-      maxPayload
+      maxPayload,
+      maxBufferedChunks,
+      maxFragments
     );
 
     this._sender = new Sender(socket, this._extensions);
@@ -411,11 +417,17 @@
  * @param {Number} options.handshakeTimeout Timeout in milliseconds for the handshake request
  * @param {Number} options.protocolVersion Value of the `Sec-WebSocket-Version` header
  * @param {String} options.origin Value of the `Origin` or `Sec-WebSocket-Origin` header
+ * @param {Number} [options.maxBufferedChunks=1048576] The maximum number of
+ *     buffered data chunks
+ * @param {Number} [options.maxFragments=131072] The maximum number of message
+ *     fragments
  * @private
  */
 function initAsClient (address, protocols, options) {
   options = Object.assign({
     protocolVersion: protocolVersions[1],
+    maxBufferedChunks: 1024 * 1024,
+    maxFragments: 128 * 1024,
     perMessageDeflate: true
   }, options, {
     createConnection: undefined,
@@ -599,7 +611,13 @@
       }
     }
 
-    this.setSocket(socket, head, 0);
+    this.setSocket(
+      socket,
+      head,
+      0,
+      options.maxBufferedChunks,
+      options.maxFragments
+    );
   });
 }
 
